service/gcs/bridge/bridge.go

```diff
@@ -14,385 +16,563 @@ import (
-			if err != nil {
-				logrus.Error(err)
-			}
-		case prot.ComputeSystemSignalProcessV1:
-			logrus.Info("received from HCS: ComputeSystemSignalProcessV1")
-			response, err = b.signalProcess(message)
-			if err != nil {
-				logrus.Error(err)
-			}
-		case prot.ComputeSystemGetPropertiesV1:
-			logrus.Info("received from HCS: ComputeSystemGetPropertiesV1")
-			response, err = b.listProcesses(message)
-			if err != nil {
-				logrus.Error(err)
+	var h Handler
+	var ok bool
+	if h, ok = mux.m[r.Header.Type]; !ok {
+		return NotSupportedHandler()
+	}
+
+	return h
+}
+
+// ServeMsg dispatches the request to the handler whose
+// type matches the request type
+func (mux *Mux) ServeMsg(w ResponseWriter, r *Request) {
+	h := mux.Handler(r)
+	h.ServeMsg(w, r)
+}
+
+// Request is the bridge request that has been sent
+type Request struct {
+	Header  *prot.MessageHeader
+	Message []byte
+}
+
+// ResponseWriter is the dispatcher used to construct the Bridge response
+type ResponseWriter interface {
+	// Header is the request header that was requested
+	Header() *prot.MessageHeader
+	// Write a successful response message.
+	Write(interface{})
+	// Error writes the provide error as a response to the message.
+	Error(error)
+}
+
+type bridgeResponse struct {
+	header   *prot.MessageHeader
+	response interface{}
+}
+
+type requestResponseWriter struct {
+	header      *prot.MessageHeader
+	respChan    chan bridgeResponse
+	respWritten bool
+}
+
+func (w *requestResponseWriter) Header() *prot.MessageHeader {
+	return w.header
+}
+
+func (w *requestResponseWriter) Write(r interface{}) {
+	w.respChan <- bridgeResponse{header: w.header, response: r}
+	w.respWritten = true
+}
+
+func (w *requestResponseWriter) Error(err error) {
+	resp := &prot.MessageResponseBase{}
+	setErrorForResponseBase(resp, err)
+	w.Write(resp)
+}
+
+// Bridge defines the bridge client in the GCS. It acts in many
+// ways analogous to go's `http` package and multiplexer.
+//
+// It has two fundamentally different dispatch options:
+//
+// 1. Request/Response where by using the `Handler` a request
+//    of a given type will be dispatched to the apprpriate handler
+//    and an appropriate `ResponseWriter` will respond to exactly
+//    that request that caused the dispatch.
+//
+// 2. `PublishNotification` where by a notification that was not initiated
+//    by a request from any client can be written to the bridge at any time
+//    in any order.
+type Bridge struct {
+	// Transport is the transport interface used by the bridge.
+	Transport transport.Transport
+
+	// Handler to invoke when messages are received
+	Handler Handler
+
+	// commandConn is the Connection the bridge receives commands (such as
+	// ComputeSystemCreate) over.
+	commandConn transport.Connection
+
+	responseChan chan bridgeResponse
+
+	// Core - TODO: Remove this and use the mux!
+	coreint core.Core
+
+	// testing hook to close the bridge ListenAndServe() method
+	quitChan chan bool
```

Comment:
Dont think it matters now but if we used `signal.Notify` for `ctrl+c` we could actually gracefully shut down the bridge listener in the future